aten/src/ATen/core/ivalue_inl.h

```diff
@@ -385,25 +377,46 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
         "Attempting to mark a completed Future as complete again. Note that "
         "a Future can only be marked completed once.");
 
+    // Start by performing all steps that can throw, before setting any field.
+    std::vector<std::reference_wrapper<const at::DataPtr>> actualDataPtrs;
+    std::vector<c10::Device> usedDevices;
     try {
-      preMarkCompletedHook(value, std::move(data_ptrs));
+      // FIXME We should always extract DataPtrs, in order to catch the case of
+      // users using CUDA values but forgetting to set devices, which currently
+      // leads to a silent synchronization/correctness issue. However, as this
+      // might worsen perf in CPU-only cases, we should only do after careful
+      // benchmarks.
+      if (impl_.type() != c10::kCPU) {
+        actualDataPtrs =
+            data_ptrs.has_value() ? std::move(*data_ptrs) : extractDataPtrs(value);
+        usedDevices = getDevicesOfDataPtrs(impl_, actualDataPtrs);
+        ensureIsSubsetOfDevices(usedDevices, devices_);
+      }
     } catch (const std::exception&) {
       setErrorInternal(std::current_exception(), lock);
       return;
     }
 
-    // Only set value_ and completed_ flag once preMarkCompletedHook has
-    // returned successfully to allow for proper error propagation.
+    // Only set value_ and completed_ flag once all checks and preparation steps
+    // have returned successfully to allow for proper error propagation.
     value_ = std::move(value);
     completed_ = true;
 
+    currentDevice_ = impl_.getDevice();
```

Comment:
I guess the main benefits of doing this here is that we can now avoid wrapping the callbacks?

Reply:
My main reason was to avoid too much branching (i.e., all those `if (impl_ == nullptr)` checks) which adds complexity to the code, hurts readability and makes testing less effective (more paths and combinations of paths to cover).